core/rawdb/accessors_chain.go

```diff
@@ -79,6 +79,33 @@ func ReadAllHashes(db ethdb.Iteratee, number uint64) []common.Hash {
 	return hashes
 }
 
+// ReadAllCanonicalHashes retrieves all canonical number and hash mappings at the
+// certain chain range. Note, this method should only used in pruned light client
+// otherwise the cost can be very expensive.
+func ReadAllCanonicalHashes(db ethdb.Iteratee, from uint64, to uint64) ([]uint64, []common.Hash) {
+	var (
+		numbers []uint64
+		hashes  []common.Hash
+	)
+	it := db.NewIteratorWithPrefix(headerPrefix)
+	defer it.Release()
+
+	for it.Next() {
+		if key := it.Key(); len(key) == len(headerPrefix)+8+1 && bytes.Equal(key[len(key)-1:], headerHashSuffix) {
+			n := binary.BigEndian.Uint64(key[len(headerPrefix) : len(headerPrefix)+8])
+			if n < from {
```

Comment:
Does not matter a lot if `from` is 1 but it would be nicer to encode `from` and append it to `headerPrefix` and start iterating from there.